### PR TITLE
Add initial dns controller

### PR DESCRIPTION
**Pull request by balooo**
This story is not complete.  This has the bulk of the changes required to calcualte the 
route53 change requests.

### CODE REVIEW

**Comment by jsravn** [comment]
Looks like wip still (commented out code, etc.)? I've updated title.

File: cmd/feed-dns/main.go [hunk 1]
```diff
@@ -36,17 +33,20 @@ func main() {
...
+	//cmd.AddSignalHandler(updater)
+
+	//err := updater.Start()
+	//if err != nil {
+	//	log.Error("Error while starting controller: ", err)
+	//	os.Exit(-1)
+	//}
 
 	select {}
 }
```
Comment (jsravn) [inline cmd/feed-dns/main.go]: This should probably be checked in a separate PR when it's finished.

File: controller/controller_test.go [hunk 2]
```diff
@@ -340,6 +340,7 @@ func TestUpdaterIsUpdatedOnK8sUpdates(t *testing.T) {
 				Path:           ingressPath,
 				ServiceAddress: serviceIP,
 				ServicePort:    ingressSvcPort,
+				FrontendName:   "internal-fronted",
```
Comment (jsravn) [inline controller/controller_test.go]: typo

File: controller/ingress_entry.go [hunk 3]
```diff
@@ -21,6 +21,8 @@ type IngressEntry struct {
 	ServicePort int32
 	// Allow are the ips or cidrs that are allowed to access the service.
 	Allow []string
+	// FrontendTag is the tag used to determine the frontend elb the host should alias
```
Comment (jsravn) [inline controller/ingress_entry.go]: Should be "FrontendName" in comment.

File: dns/dns_updater.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+)
+
+type updater struct {
+	dns          r53.Route53Client
+	elb          elb.ELB
+	elbLabelName string
+}
+
+// New creates an updater for dns
+func New(dnsClient r53.Route53Client, awsElb elb.ELB, elbLabelName string) controller.Updater {
```
Comment (jsravn) [inline dns/dns_updater.go]: Any reason we don't create the route53 client internally here? This is bleeding the aws dependency out of the dns package.

File: dns/dns_updater.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	return nil
+}
+
+func (u *updater) Health() error {
+
+	return nil
+}
+
+func (u *updater) Update(update controller.IngressUpdate) error {
+	frontEnds, err := elb.FindFrontEndElbs(u.elb, u.elbLabelName)
```
Comment (jsravn) [inline dns/dns_updater.go]: This needs to be cached or updated asynchronously. Like `feed-ingress` we should probably just query it at startup, with an `expected-frontends` parameter to ensure we find everything we expect. Otherwise this may hammer aws.

File: dns/dns_updater.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/sky-uk/feed/controller"
+	"github.com/sky-uk/feed/dns/r53"
+	"github.com/sky-uk/feed/elb"
+	"k8s.io/heapster/Godeps/_workspace/src/github.com/coreos/fleet/log"
```
Comment (jsravn) [inline dns/dns_updater.go]: wrong import

File: dns/dns_updater.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+		frontEnd, exists := frontEnds[ingressEntry.FrontendName]
+		if !exists {
+			return nil, fmt.Errorf("unable to find front end loadd balancer with name: %v", ingressEntry.FrontendName)
+		}
+		changes = append(changes, newChange("UPSERT", ingressEntry.Host, frontEnd.DNSName, frontEnd.HostedZoneID))
+	}
+
+	for _, recordSet := range aRecords {
+		if _, contains := hostToIngresEntry[*recordSet.Name]; !contains {
+			changes = append(changes, newChange("DELETE", *recordSet.Name, *recordSet.AliasTarget.DNSName, *recordSet.AliasTarget.HostedZoneId))
```
Comment (jsravn) [inline dns/dns_updater.go]: Mind breaking up the long lines? Try to keep <120 chars.

File: dns/dns_updater.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+		}
+	}
+
+	log.Infof("calculated changes to dns: %v", changes)
```
Comment (jsravn) [inline dns/dns_updater.go]: This is going to print out the entire struct, it's going to be hard to interpret. What about printing out just a list of the hostnames added or something like that? Or a count.

File: dns/r53/client.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+import (
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/route53"
+)
+
+const (
+	hostedZoneID = "Z2KHQXXHNXGTXW"
```
Comment (jsravn) [inline dns/r53/client.go]: Needs to be configurable.

File: dns/r53/client.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+	return nil
+}
+
+// GetARecords gets a list of A Records from aws.  (Currently uses a hardcoded hostedZoneID but this will be parameterized)
+func (dns *Route53Client) GetARecords() []*route53.ResourceRecordSet {
+
+	recordSetsOutput, err := dns.client.ListResourceRecordSets(&route53.ListResourceRecordSetsInput{
+		HostedZoneId: aws.String(hostedZoneID),
+	})
```
Comment (jsravn) [inline dns/r53/client.go]: This is paginated so you need to get all pages.

File: dns/r53/client.go [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+	hostedZoneID = "Z2KHQXXHNXGTXW"
+)
+
+// Route53Client enables interaction with aws route53
+type Route53Client struct {
+	client route53.Route53
+}
+
+// New creates a route53 client used to interact with aws
+func New() Route53Client {
```
Comment (jsravn) [inline dns/r53/client.go]: Did we decide not to have any test coverage for the client?

File: dns/r53/client.go [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+// UpdateRecordSets updates records in aws based on the change list.
+func (dns *Route53Client) UpdateRecordSets(changes []*route53.Change) error {
+
+	recordSetsInput := &route53.ChangeResourceRecordSetsInput{
+		HostedZoneId: aws.String(hostedZoneID),
+		ChangeBatch: &route53.ChangeBatch{
+			Changes: changes,
+		},
+	}
```
Comment (jsravn) [inline dns/r53/client.go]: Does this have some sort of limit on number of changes we can make in a single call?

File: dns/dns_updater.go (hunk 4 above)
Comment (chbatey) [inline dns/dns_updater.go]: agreed, put it inside

File: dns/dns_updater.go (hunk 5 above)
Comment (chbatey) [inline dns/dns_updater.go]: For the initial version I will do it once on startup

File: dns/dns_updater.go (hunk 8 above)
Comment (chbatey) [inline dns/dns_updater.go]: I am finding the full output very useful during development / debugging. I will remove this in a week :)

File: dns/r53/client.go (hunk 9 above)
Comment (chbatey) [inline dns/r53/client.go]: El done!

File: dns/r53/client.go (hunk 11 above)
Comment (chbatey) [inline dns/r53/client.go]: I will add tests at some point

File: dns/r53/client.go (hunk 12 above)
Comment (chbatey) [inline dns/r53/client.go]: 100

I will add it to the story to deal with a greater number and do in a different PR as this PR is humongous

File: dns/r53/client.go (hunk 10 above)
Comment (chbatey) [inline dns/r53/client.go]: Have added it as a task to the story

**Comment by chbatey** [comment]
I've made this runnable (removed commented out code / fixed main) and addressed all of james's comments b by either fixing it or creating a future task
